### PR TITLE
Fix numerous signed unsigned mismatch warnings.

### PR DESCRIPTION
**Pull request by jgiszczak**
## Change Description

Fix numerous signed unsigned mismatch warnings.

## Consensus Changes

N/A

## API Changes

N/A

## Documentation Additions

N/A

### CODE REVIEW

File: libraries/chain/wasm_interface.cpp [hunk 1]
```diff
@@ -1266,7 +1266,7 @@ class memory_api : public context_aware_api {
       :context_aware_api(ctx,true){}
 
       char* memcpy( array_ptr<char> dest, array_ptr<const char> src, size_t length) {
-         EOS_ASSERT((std::abs((ptrdiff_t)dest.value - (ptrdiff_t)src.value)) >= length,
+         EOS_ASSERT((size_t)(std::abs((ptrdiff_t)dest.value - (ptrdiff_t)src.value)) >= length,
```
Comment (heifner) [inline libraries/chain/wasm_interface.cpp]: This is consensus code but the abs should always return a positive value, so this change should have no ill-affect.

File: libraries/testing/tester.cpp [hunk 2]
```diff
@@ -797,7 +797,7 @@ namespace eosio { namespace testing {
          return other.sync_with(*this);
 
       auto sync_dbs = [](base_tester& a, base_tester& b) {
-         for( int i = 1; i <= a.control->head_block_num(); ++i ) {
+         for( unsigned int i = 1; i <= a.control->head_block_num(); ++i ) {
             auto block = a.control->fetch_block_by_number(i);
```
Comment (heifner) [inline libraries/testing/tester.cpp]: head_block_num() returns a `uint32_t`, would like this to match.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 3]
```diff
@@ -880,7 +880,7 @@ bool chain_plugin::import_reversible_blocks( const fc::path& reversible_dir,
    reversible_blocks.open( reversible_blocks_file.generic_string().c_str(), std::ios::in | std::ios::binary );
 
    reversible_blocks.seekg( 0, std::ios::end );
-   uint64_t end_pos = reversible_blocks.tellg();
+   int64_t end_pos = reversible_blocks.tellg();
    reversible_blocks.seekg( 0 );
```
Comment (heifner) [inline plugins/chain_plugin/chain_plugin.cpp]: `tellg()` returns a `basic_istream<>::pos_type` I think I would rather see `auto` in this case.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 4]
```diff
@@ -1585,7 +1585,7 @@ static void push_recurse(read_write* rw, int index, const std::shared_ptr<read_w
          results->emplace_back( r );
       }
 
-      int next_index = index + 1;
+      unsigned int next_index = index + 1;
       if (next_index < params->size()) {
```
Comment (heifner) [inline plugins/chain_plugin/chain_plugin.cpp]: `size_t`

File: plugins/txn_test_gen_plugin/txn_test_gen_plugin.cpp [hunk 5]
```diff
@@ -98,7 +98,7 @@ struct txn_test_gen_plugin_impl {
       chain_plugin& cp = app().get_plugin<chain_plugin>();
 
       const int overlap = 20;
-      int end = std::min(index + overlap, trxs->size());
+      unsigned int end = std::min(index + overlap, trxs->size());
       _remain = end - index;
```
Comment (heifner) [inline plugins/txn_test_gen_plugin/txn_test_gen_plugin.cpp]: `size_t`

File: programs/cleos/main.cpp [hunk 6]
```diff
@@ -1850,7 +1850,7 @@ void get_account( const string& accountName, const string& coresym, bool json_fo
             auto& prods = obj["producers"].get_array();
             std::cout << "producers:";
             if ( !prods.empty() ) {
-               for ( int i = 0; i < prods.size(); ++i ) {
+               for ( unsigned int i = 0; i < prods.size(); ++i ) {
                   if ( i%3 == 0 ) {
```
Comment (heifner) [inline programs/cleos/main.cpp]: `size_t`

File: programs/eosio-launcher/main.cpp [hunk 7]
```diff
@@ -850,7 +850,7 @@ launcher_def::bind_nodes () {
       cerr << "Unable to allocate producers due to insufficient prod_nodes = " << prod_nodes << "\n";
       exit (10);
    }
-   int non_bios = prod_nodes - 1;
+   unsigned int non_bios = prod_nodes - 1;
    int per_node = producers / non_bios;
```
Comment (heifner) [inline programs/eosio-launcher/main.cpp]: size_t

File: unittests/api_tests.cpp [hunk 8]
```diff
@@ -201,7 +201,7 @@ transaction_trace_ptr CallAction(TESTER& test, T ac, const vector<account_name>&
 
    auto pl = vector<permission_level>{{scope[0], config::active_name}};
    if (scope.size() > 1)
-      for (int i = 1; i < scope.size(); i++)
+      for (unsigned int i = 1; i < scope.size(); i++)
          pl.push_back({scope[i], config::active_name});
```
Comment (heifner) [inline unittests/api_tests.cpp]: size_t

File: unittests/forked_tests.cpp [hunk 9]
```diff
@@ -125,7 +125,7 @@ BOOST_AUTO_TEST_CASE( fork_with_bad_block ) try {
       BOOST_TEST_CONTEXT("Testing Fork: " << i) {
          const auto& fork = forks.at(i);
          // push the fork to the original node
-         for (int fidx = 0; fidx < fork.blocks.size() - 1; fidx++) {
+         for (unsigned int fidx = 0; fidx < fork.blocks.size() - 1; fidx++) {
             const auto& b = fork.blocks.at(fidx);
```
Comment (heifner) [inline unittests/forked_tests.cpp]: size_t

File: unittests/forked_tests.cpp [hunk 10]
```diff
@@ -342,7 +342,7 @@ BOOST_AUTO_TEST_CASE( prune_remove_branch ) try {
    BOOST_REQUIRE_EQUAL(73, c2.control->head_block_num());
    
    // push fork from c2 => c
-   int p = fork_num;
+   unsigned int p = fork_num;
```
Comment (heifner) [inline unittests/forked_tests.cpp]: `uint32_t`

File: unittests/message_buffer_tests.cpp [hunk 11]
```diff
@@ -227,7 +227,7 @@ BOOST_AUTO_TEST_CASE(message_buffer_write_ptr_to_end)
       BOOST_CHECK_EQUAL(mb.write_index().second, 0);
 
       char* write_ptr = mb.write_ptr();
-      for (char ind = 0; ind < small; ind++) {
+      for (unsigned char ind = 0; ind < small; ind++) {
```
Comment (heifner) [inline unittests/message_buffer_tests.cpp]: uint32_t

File: unittests/misc_tests.cpp [hunk 12]
```diff
@@ -542,7 +542,7 @@ BOOST_AUTO_TEST_CASE(alphabetic_sort)
     tmp.push_back(str);
   }
 
-  for(int i = 0; i < words.size(); ++i ) {
+  for(unsigned int i = 0; i < words.size(); ++i ) {
     BOOST_TEST(tmp[i] == words[i]);
```
Comment (heifner) [inline unittests/misc_tests.cpp]: size_t

File: unittests/resource_limits_test.cpp [hunk 13]
```diff
@@ -216,7 +216,7 @@ BOOST_AUTO_TEST_SUITE(resource_limits_test)
       const uint64_t increment = 1000;
       const uint64_t expected_iterations = config::default_max_block_cpu_usage / increment;
 
-      for (int idx = 0; idx < expected_iterations; idx++) {
+      for (unsigned int idx = 0; idx < expected_iterations; idx++) {
          add_transaction_usage({account}, increment, 0, 0);
```
Comment (heifner) [inline unittests/resource_limits_test.cpp]: uint64_t

File: unittests/resource_limits_test.cpp [hunk 14]
```diff
@@ -233,7 +233,7 @@ BOOST_AUTO_TEST_SUITE(resource_limits_test)
       const uint64_t increment = 1000;
       const uint64_t expected_iterations = config::default_max_block_net_usage / increment;
 
-      for (int idx = 0; idx < expected_iterations; idx++) {
+      for (unsigned int idx = 0; idx < expected_iterations; idx++) {
          add_transaction_usage({account}, 0, increment, 0);
```
Comment (heifner) [inline unittests/resource_limits_test.cpp]: uint64_t

File: unittests/resource_limits_test.cpp [hunk 15]
```diff
@@ -252,7 +252,7 @@ BOOST_AUTO_TEST_SUITE(resource_limits_test)
       set_account_limits(account, limit, -1, -1 );
       process_account_limit_updates();
 
-      for (int idx = 0; idx < expected_iterations - 1; idx++) {
+      for (unsigned int idx = 0; idx < expected_iterations - 1; idx++) {
          add_pending_ram_usage(account, increment);
```
Comment (heifner) [inline unittests/resource_limits_test.cpp]: uint64_t

File: unittests/special_accounts_tests.cpp [hunk 16]
```diff
@@ -62,7 +62,7 @@ BOOST_FIXTURE_TEST_CASE(accounts_exists, tester)
       }
 
       std::vector<account_name> diff;
-      for (int i = 0; i < std::max(active_auth.size(), active_producers.producers.size()); ++i) {
+      for (unsigned int i = 0; i < std::max(active_auth.size(), active_producers.producers.size()); ++i) {
          account_name n1 = i < active_auth.size() ? active_auth[i] : (account_name)0;
```
Comment (heifner) [inline unittests/special_accounts_tests.cpp]: size_t

**Comment by jgiszczak** [comment]
Per the C standard §6.3.1.8, when comparing signed and unsigned, the compiler is required to convert signed to unsigned and then the comparison is performed.

**Pull request merged**